executor/distsql.go

```diff
@@ -623,15 +623,40 @@ func (e *IndexLookUpExecutor) startWorkers(ctx context.Context, initBatchSize in
 	return nil
 }
 
-func (e *IndexLookUpExecutor) hasExtralPidCol() bool {
-	return e.index.Global || (e.partitionTableMode && len(e.byItems) > 0)
+func (e *IndexLookUpExecutor) needPartitionHandle(tp getHandleType) (bool, error) {
+	var col *expression.Column
+	var needPartitionHandle, ret bool
+	if tp == getHandleFromIndex {
+		cols := e.idxPlans[0].Schema().Columns
+		outputOffsets := e.dagPB.OutputOffsets
+		col = cols[outputOffsets[len(outputOffsets)-1]]
+		needPartitionHandle = e.index.Global || (e.partitionTableMode && e.keepOrder)
+		ret = col.ID == model.ExtraPhysTblID || col.ID == model.ExtraPidColID
+	} else {
+		cols := e.tblPlans[0].Schema().Columns
+		outputOffsets := e.tableRequest.OutputOffsets
+		col = cols[outputOffsets[len(outputOffsets)-1]]
+
+		needPartitionHandle = e.index.Global || (e.partitionTableMode && e.keepOrder)
+		// no ExtraPidColID here, because tableScan shouldn't contain them.
+		ret = col.ID == model.ExtraPhysTblID
+	}
+
+	// TODO: fix global index related bugs later
+	if needPartitionHandle != ret && !e.index.Global {
+		return ret, errors.Errorf("Internal error, needPartitionHandle(%t) != ret(%t), tp(%d)", needPartitionHandle, ret, tp)
+	}
+	return ret, nil
 }
 
 func (e *IndexLookUpExecutor) isCommonHandle() bool {
 	return !(len(e.handleCols) == 1 && e.handleCols[0].ID == model.ExtraHandleID) && e.table.Meta() != nil && e.table.Meta().IsCommonHandle
 }
 
 func (e *IndexLookUpExecutor) getRetTpsForIndexReader() []*types.FieldType {
+	if e.checkIndexValue != nil {
```

Comment:
Is this a bug fix?

Reply:
No, just move it to the front, make code more clean